docs/proposals/local-registries.md

```diff
@@ -0,0 +1,165 @@
+
+### Non-Goals
+
+- Automate how pull/push secrets can be added to a `Build` object.
+- Support secure image pull/push if a local registry uses self-signed certificates.
+
+## Proposal
+
+Application developers will be able to specify `local: true|false` if an image reference points
+to a cluster-local image registry.
+
+The build controller/operator will watch the global "LocalRegistryHosting" `ConfigMap` to check if
+a local image registry is declared. If declared, it will infer the local registry host information
+from what is provided in the `ConfigMap`.
+
+When a `BuildRun` is instantiated, the image specs that declare `local: true` are mutated to inject
+the local image registry host.
+
+### User Stories [optional]
+
+As an application developer
+I would like to declare that images in builds reference a local image registry
+So that I do not have to provide host information in the pull spec
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Obtaining local registry information
+
+First, clusters which deploy a local image registry must provide the releavant host info by
+publishing it in the `local-registry-hosting`
+[ConfigMap](https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/generic/1755-communicating-a-local-registry#the-local-registry-hosting-configmap).
+The build operator/controller must have RBAC which allows it to read the `ConfigMap` and obtain the
+registry host information. For `Build` and `BuildRun` objects, the `HostFromClusterNetwork` value
+should be used as the hostname.
+
+If the `local-registry-hosting` ConfigMap is created, updated, or deleted, the build controller
+and/or operator needs to react accordingly.
+
+#### Declaring an image as local
+
+`Image` references in the API will be updated to have the `local` field. An example using the 
+`output` spec:
+
+```yaml
+spec:
+  output:
+    image: mynamespace/myapp:latest
+    local: true
+```
+
+When an image is declared as `local`, the build controller prepends the local image registry
+hostname to the provided image reference.
+
+### Risks and Mitigations
+
+App developers use `local: true` on clusters that don't have a local registry defined, end up
+pulling images from `docker.io` or an image registry in container runtime's unqualified search
+path.
+
+Local image registry configurations can be altered while the build controller is running.
+
+## Design Details
+
+### Test Plan
+
+1. Deploy the build operator/controller on a cluster that populates the `local-registry-hosting`
+   ConfigMap OR as cluster-admin, manually populate the `local-registry-hosting` ConfigMap.
+2. Create a `Build` and/or `BuildRun` which references a local image.
+3. Ensure that a build can push an image referencing the local image spec.
+
+### Graduation Criteria
+
+**Dev Preview** - we can simply define an environment variable to set the local registry hostname.
```

Comment:
When looking at your [code](https://github.com/redhat-developer/build/compare/master...adambkaplan:local-registries-impl?expand=1#diff-c8bae8bebf9ceb3609b947afd6c4af05R193) it seems this part is not yet addressed. Ideally we want to be able to set this ENV variable per Build. This will imply one new API endpoint for the `Build`/`BuildRun`, plus the one for the `spec.output.local`. Also, to consider that at this point , the local registry endpoint can be addressed already by just modifying the `spec.output.url`

Reply:
@qu1queee the environment variable I referenced is [here](https://github.com/redhat-developer/build/compare/master...adambkaplan:local-registries-impl?expand=1#diff-091ace02b49a46f55f2586cfec89a160R24).

> Ideally we want to be able to set this ENV variable per Build. This will imply one new API endpoint for the Build/BuildRun, plus the one for the spec.output.local

I disagree. A fundamental assumption of this proposal is:

1. There is one local registry per cluster.
2. The hostname of the local registry can vary across distributions.

Dropping the hostname allows `Build` and `BuildRun` objects to be portable. Therefore, we do not want the hostname to be changed/overrode in the `Build`/`BuildRun` API. As you alluded to, this is no different than changing the raw URL.

Reply:
make sense, thanks